// @sudoku/stores/explore.js
// src/node_modules/@sudoku/stores/explore.js
import { writable, derived, get } from 'svelte/store';
import { cloneSudoku, hasConflict as checkSudokuConflict } from '@sudoku/sudoku';
import { history } from './history';
import { userGrid } from './grid';

function createExplore() {
	// 独立的状态store
	const isExploring = writable(false);
	const exploreConflict = writable(false);
	const failedPaths = writable(new Set());
	
	// 内部状态，不对外暴露
	let startSnapshot = null;
	let startHistoryLength = 0;

	// 订阅userGrid变化，自动检查探索冲突
	userGrid.subscribe(() => {
		if (get(isExploring)) {
			checkConflict();
		}
	});

	// 统一导出所有状态的derived store
	const exploreState = derived(
		[isExploring, exploreConflict, failedPaths],
		([$isExploring, $exploreConflict, $failedPaths]) => ({
			isExploring: $isExploring,
			hasConflict: $exploreConflict,
			failedPaths: $failedPaths
		})
	);

	/**
	 * 检查当前探索是否有冲突
	 * 同时检查是否在失败路径中
	 */
	function checkConflict() {
		const currentGrid = get(userGrid);
		const conflict = checkSudokuConflict(currentGrid);
		
		// 检查是否在已知失败路径中
		const gridKey = JSON.stringify(currentGrid);
		const isFailedPath = get(failedPaths).has(gridKey);

		exploreConflict.set(conflict || isFailedPath);
		return conflict || isFailedPath;
	}

	return {
		subscribe: exploreState.subscribe,

		/**
		 * 进入探索模式
		 * 保存当前状态作为探索起点
		 */
		enterExplore() {
			const currentGrid = get(userGrid);
			startSnapshot = cloneSudoku(currentGrid);
			startHistoryLength = get(history.past).length;

			isExploring.set(true);
			exploreConflict.set(false);
		},

		/**
		 * 提交探索结果
		 * 将探索过程中的所有操作永久合并入主历史
		 */
		commitExplore() {
			// 检查是否有冲突
			if (get(exploreConflict)) {
				throw new Error('无法提交有冲突的探索结果');
			}

			// 提交成功，清除探索状态
			isExploring.set(false);
			exploreConflict.set(false);
			startSnapshot = null;
			startHistoryLength = 0;
		},

		/**
		 * 放弃探索，精确回滚到探索起点
		 */
		abandonExplore() {
			const currentHistoryLength = get(history.past).length;
			// 精确计算需要撤销的步数
			const stepsToUndo = currentHistoryLength - startHistoryLength;

			// 撤销探索期间的所有操作
			for (let i = 0; i < stepsToUndo; i++) {
				userGrid.undoGrid();
			}

			// 清空redo栈，防止回到已放弃的路径
			history.future.set([]);

			// 记录失败路径（记录探索起点）
			const failedGrid = get(userGrid);
			failedPaths.update(set => {
				const newSet = new Set(set);
				newSet.add(JSON.stringify(failedGrid));
				return newSet;
			});

			// 清除探索状态
			isExploring.set(false);
			exploreConflict.set(false);
			startSnapshot = null;
			startHistoryLength = 0;
		},

		/**
		 * 重置失败路径记忆
		 */
		resetFailedPaths() {
			failedPaths.set(new Set());
		}
	};
}

export const explore = createExplore();
