// @sudoku/stores/hints.js
// src/node_modules/@sudoku/stores/hints.js
import { settings } from './settings';
import { writable, get } from 'svelte/store';
import { getNextHint, getCandidates } from '@sudoku/sudoku';
import { userGrid } from './grid';
import { cursor } from './cursor';

export const usedHints = writable(0);
export const currentHint = writable(null);

function createHints() {
	let defaultHints = Infinity;

	const hints = writable(Infinity);

	settings.subscribe(($settings) => {
		if ($settings.hintsLimited) {
			defaultHints = $settings.hints;
			hints.update($hints => {
				if ($hints > $settings.hints) return $settings.hints;

				return $hints;
			})
		} else {
			defaultHints = Infinity;
			hints.set(Infinity);
		}
	});

	return {
		subscribe: hints.subscribe,

		getCandidatesAtCursor() {
			const $cursor = get(cursor);
			const $grid = get(userGrid);
			if ($cursor.x === null || $cursor.y === null) return [];
			return getCandidates($grid, $cursor.x, $cursor.y);
		},

		getNextStepHint() {
			const $grid = get(userGrid);
			return getNextHint($grid);
		},

		useHint() {
			hints.update($hints => {
				if ($hints > 0) {
					usedHints.update($usedHints => $usedHints + 1);
					return $hints - 1;
				}

				return 0;
			});
		},

		reset() {
			hints.set(defaultHints);
			usedHints.set(0);
			currentHint.set(null);
		}
	};
}

export const hints = createHints();
