// @sudoku/stores/history.js
// src/node_modules/@sudoku/stores/history.js
import { writable } from 'svelte/store';
import { cloneSudoku } from '@sudoku/sudoku';

function createHistory() {
	const past = writable([]);
	const future = writable([]);

	return {
		past,
		future,

		/**
		 * 操作前调用，将当前状态压入历史栈
		 * 同时清空redo栈
		 */
		pushState(grid) {
			past.update(p => [...p, cloneSudoku(grid)]);
			future.set([]);
		},

		/**
		 * 撤销操作
		 * 将当前状态存入redo栈，返回undo后的状态
		 */
		undo(currentGrid) {
			let result = null;
			past.update(p => {
				if (p.length > 0) {
					// 将当前状态存入future栈
					future.update(f => [...f, cloneSudoku(currentGrid)]);
					// 返回past栈顶元素
					result = p[p.length - 1];
					// 移除栈顶元素
					return p.slice(0, -1);
				}
				return p;
			});
			return result;
		},

		/**
		 * 重做操作
		 * 将当前状态存入undo栈，返回redo后的状态
		 */
		redo(currentGrid) {
			let result = null;
			future.update(f => {
				if (f.length > 0) {
					// 将当前状态存入past栈
					past.update(p => [...p, cloneSudoku(currentGrid)]);
					// 返回future栈顶元素
					result = f[f.length - 1];
					// 移除栈顶元素
					return f.slice(0, -1);
				}
				return f;
			});
			return result;
		},

		/**
		 * 清空历史记录
		 * 新游戏时调用
		 */
		reset() {
			past.set([]);
			future.set([]);
		}
	};
}

export const history = createHistory();
