// @sudoku/stores/history.js
// src/node_modules/@sudoku/stores/history.js
import { writable } from 'svelte/store';
import { cloneSudoku } from '@sudoku/sudoku';

function createHistory() {
	const past = writable([]);
	const future = writable([]);

	let presentSnapshot = null;

	return {
		// 保存当前状态（在每次 userGrid 改变前调用）
		pushState(grid) {
			past.update(p => {
				p.push(presentSnapshot);
				return p;
			});
			presentSnapshot = cloneSudoku(grid);
			future.set([]); // 清空未来
		},

		// 撤销
		undo(currentGrid) {
			let result = null;
			past.update(p => {
				if (p.length > 0) {
					future.update(f => {
						f.push(presentSnapshot);
						return f;
					});
					presentSnapshot = p.pop();
					result = presentSnapshot;
				}
				return p;
			});
			return result;
		},

		// 重做
		redo(currentGrid) {
			let result = null;
			future.update(f => {
				if (f.length > 0) {
					past.update(p => {
						p.push(presentSnapshot);
						return p;
					});
					presentSnapshot = f.pop();
					result = presentSnapshot;
				}
				return f;
			});
			return result;
		},

		// 重置历史
		reset(initialGrid) {
			past.set([]);
			future.set([]);
			presentSnapshot = cloneSudoku(initialGrid);
		}
	};
}

export const history = createHistory();
