// @sudoku/sudoku.js
// src/node_modules/@sudoku/sudoku.js

import solve from '@mattflow/sudoku-solver';
import { BOX_SIZE, GRID_LENGTH, SUDOKU_SIZE, GRID_COORDS } from '@sudoku/constants';
import { getSudoku } from 'fake-sudoku-puzzle-generator';

/**
 * @param {('veryeasy' | 'easy' | 'medium' | 'hard')} difficulty
 * @returns {number[][]}
 */
export function generateSudoku(difficulty = 'easy') {
	const sudoku = getSudoku(difficulty);

	for (let row = 0; row < SUDOKU_SIZE; row++) {
		for (let col = 0; col < SUDOKU_SIZE; col++) {
			if (sudoku[row][col] === null) sudoku[row][col] = 0;
		}
	}

	return sudoku;
}

/**
 * @param {number[][]} sudoku
 */
export function solveSudoku(sudoku) {
	let grid = [
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
	];

	const solution = solve(sudoku.flat().join(''), {
		outputArray: true,
		hintCheck: false
	});

	for (let cell = 0; cell < GRID_LENGTH; cell++) {
		const [row, col] = GRID_COORDS[cell];
		grid[row][col] = solution[cell];
	}

	return grid;
}

/**
 * @param {number[][]} sudoku
 */
export function printSudoku(sudoku) {
	let out = '╔═══════╤═══════╤═══════╗\n';

	for (let row = 0; row < SUDOKU_SIZE; row++) {
		if (row !== 0 && row % BOX_SIZE === 0) {
			out += '╟───────┼───────┼───────╢\n';
		}

		for (let col = 0; col < SUDOKU_SIZE; col++) {
			if (col === 0) {
				out += '║ ';
			} else if (col % BOX_SIZE === 0) {
				out += '│ ';
			}

			out += (sudoku[row][col] === 0 ? '·' : sudoku[row][col]) + ' ';

			if (col === SUDOKU_SIZE - 1) {
				out += '║';
			}
		}

		out += '\n';
	}

	out += '╚═══════╧═══════╧═══════╝';

	console.log(out);
}

/**
 * 获取某个单元格的候选数集合
 * @param {number[][]} sudoku - 当前棋盘
 * @param {number} x - 列坐标
 * @param {number} y - 行坐标
 * @returns {number[]} 候选数数组
 */
export function getCandidates(sudoku, x, y) {
	if (sudoku[y][x] !== 0) return [];

	const candidates = new Set([1, 2, 3, 4, 5, 6, 7, 8, 9]);

	// 检查行
	for (let i = 0; i < SUDOKU_SIZE; i++) {
		candidates.delete(sudoku[y][i]);
	}

	// 检查列
	for (let i = 0; i < SUDOKU_SIZE; i++) {
		candidates.delete(sudoku[i][x]);
	}

	// 检查 3x3 宫
	const startY = Math.floor(y / BOX_SIZE) * BOX_SIZE;
	const startX = Math.floor(x / BOX_SIZE) * BOX_SIZE;
	for (let row = startY; row < startY + BOX_SIZE; row++) {
		for (let col = startX; col < startX + BOX_SIZE; col++) {
			candidates.delete(sudoku[row][col]);
		}
	}

	return Array.from(candidates);
}

/**
 * 获取下一步提示（唯一候选数的格子）
 * @param {number[][]} sudoku - 当前棋盘
 * @returns { {x: number, y: number, value: number, reason: string} | null }
 */
export function getNextHint(sudoku) {
	// 策略 1: 找唯一候选数（Naked Single）
	for (let y = 0; y < SUDOKU_SIZE; y++) {
		for (let x = 0; x < SUDOKU_SIZE; x++) {
			if (sudoku[y][x] === 0) {
				const candidates = getCandidates(sudoku, x, y);
				if (candidates.length === 1) {
					return {
						x, y,
						value: candidates[0],
						reason: '唯一候选数 (Naked Single)'
					};
				}
			}
		}
	}

	// 策略 2: 找隐式唯一数（Hidden Single）- 仅在行/列/宫中某个位置出现的数
	for (let num = 1; num <= 9; num++) {
		// 检查行
		for (let y = 0; y < SUDOKU_SIZE; y++) {
			let positions = [];
			for (let x = 0; x < SUDOKU_SIZE; x++) {
				if (sudoku[y][x] === 0 && getCandidates(sudoku, x, y).includes(num)) {
					positions.push({ x, y });
				}
			}
			if (positions.length === 1) {
				return { ...positions[0], value: num, reason: '行隐式唯一数 (Hidden Single in Row)' };
			}
		}

		// 检查列
		for (let x = 0; x < SUDOKU_SIZE; x++) {
			let positions = [];
			for (let y = 0; y < SUDOKU_SIZE; y++) {
				if (sudoku[y][x] === 0 && getCandidates(sudoku, x, y).includes(num)) {
					positions.push({ x, y });
				}
			}
			if (positions.length === 1) {
				return { ...positions[0], value: num, reason: '列隐式唯一数 (Hidden Single in Column)' };
			}
		}

		// 检查宫
		for (let boxY = 0; boxY < 3; boxY++) {
			for (let boxX = 0; boxX < 3; boxX++) {
				let positions = [];
				const startY = boxY * BOX_SIZE;
				const startX = boxX * BOX_SIZE;
				
				for (let y = startY; y < startY + BOX_SIZE; y++) {
					for (let x = startX; x < startX + BOX_SIZE; x++) {
						if (sudoku[y][x] === 0 && getCandidates(sudoku, x, y).includes(num)) {
							positions.push({ x, y });
						}
					}
				}
				
				if (positions.length === 1) {
					return { ...positions[0], value: num, reason: '宫隐式唯一数 (Hidden Single in Box)' };
				}
			}
		}
	}

	return null; // 没有可逻辑推断的下一步，需要进入探索模式
}

/**
 * 检查棋盘是否有冲突
 * @param {number[][]} sudoku - 当前棋盘
 * @returns {boolean} 是否有冲突
 */
export function hasConflict(sudoku) {
	for (let y = 0; y < SUDOKU_SIZE; y++) {
		for (let x = 0; x < SUDOKU_SIZE; x++) {
			const value = sudoku[y][x];
			if (value !== 0) {
				// 检查行
				for (let i = 0; i < SUDOKU_SIZE; i++) {
					if (i !== x && sudoku[y][i] === value) return true;
				}
				// 检查列
				for (let i = 0; i < SUDOKU_SIZE; i++) {
					if (i !== y && sudoku[i][x] === value) return true;
				}
				// 检查宫
				const startY = Math.floor(y / BOX_SIZE) * BOX_SIZE;
				const startX = Math.floor(x / BOX_SIZE) * BOX_SIZE;
				for (let row = startY; row < startY + BOX_SIZE; row++) {
					for (let col = startX; col < startX + BOX_SIZE; col++) {
						if (row !== y && col !== x && sudoku[row][col] === value) return true;
					}
				}
			}
		}
	}
	return false;
}

/**
 * 深拷贝棋盘
 * @param {number[][]} sudoku - 原棋盘
 * @returns {number[][]} 新棋盘
 */
export function cloneSudoku(sudoku) {
	return sudoku.map(row => [...row]);
}
